akka-actor/src/main/scala/akka/util/OptionVal.scala

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.util
+
+/**
+ * INTERNAL API
+ */
+private[akka] object OptionVal {
+
+  def apply[A](x: A): OptionVal[A] =
+    new OptionVal(x)
+
+  /**
+   * Represents non-existent values, `null` values.
+   */
+  val None = new OptionVal(null)
+}
+
+/**
+ * INTERNAL API
+ * Represents optional values similar to `scala.Option`, but
+ * as a value class to avoid allocations.
+ *
+ * Note that it should not be used in pattern matching to avoid
+ * allocations. That is also the reason why it's not a `case class`
+ * with `unapply`.
+ * See http://docs.scala-lang.org/overviews/core/value-classes.html#when-allocation-is-necessary
```

Comment:
Just checked, it's "not really documented" in Scala, but since 2.11 thanks to name based extraction if a class has:
`def isEmpty: Boolean` and `def get: T` then it can be used in pattern matching *without allocations*!
In this impl we'll end up double checking `!= null` but that's perhaps fine, still cheaper than allocating